### ISSUE TITLE
Deprecating a type parameter doesn't warn when the parameter is used.

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#5120]
**Describe the issue**

In some cases, a package may want one class to stop being used with generics.   
For example, Riverpod refactored `Ref<StateT>` to just `Ref`.

This change is done with:

```diff
+@optionalTypeArgs
class Ref<
+  @deprecated
  T> {}
```

The problem is, existing references to the type parameter does not cause a warning.

**To Reproduce**

```dart
@optionalTypeArgs
class Ref<@deprecated  T> {}

Ref<int>? value; 
```

**Expected behavior**

This should warn on `<int>` that the type parameter is deprecated
